pontoon/base/models.py

```diff
@@ -627,7 +635,30 @@ def for_project_locale(self, project, locale, paths=None):
             obsolete=False
         )
 
-        if paths:
+        translations = Translation.objects.filter(locale=locale)
+
+        # Filter by search parameters
+        if search:
+            q = search.get('q', None)
+            i = '' if search.get('casesensitive', None) else 'i'
+
+            entity_query = Q()  # Empty object
+            if search.get('sources', None):
+                entity_query |= Q(**{'string__%scontains' % i: q}) | Q(**{'string_plural__%scontains' % i: q})
+
+            if search.get('translations', None):
+                entity_query |= Q(translation__in=translations.filter(**{'string__%scontains' % i: q}))
+
+            if search.get('comments', None):
+                entity_query |= Q(**{'comment__%scontains' % i: q})
+
+            if search.get('keys', None):
+                entity_query |= Q(**{'key__%scontains' % i: q})
+
+            entities = entities.filter(entity_query).distinct()
+
+        # Filter by path
+        elif paths:
```

Comment:
I don't have a good solution so we can keep it for now, but having two optional parameters that are mutually exclusive is a bad code smell. I think sometime in the future we'll want to refactor this into multiple methods or something.